### PR TITLE
gen/json: Enable omitempty for optional collections

### LINKED ISSUE #330: Add `omitempty` json tag for optional list and map fields
As per `encoding/json` documentation: https://golang.org/pkg/encoding/json/#Marshal
> The "omitempty" option specifies that the field should be omitted from the encoding if the field has an empty value, defined as false, 0, a nil pointer, a nil interface value, and any empty array, slice, map, or string.

If an thrift struct definition has an optional list or map:

```
struct OptionalListField {
    1: optional list<Item> items
    2: optional map<Key, Value> dictionary
}
```

then thriftrw should add the `omitempty` json tag:

```
type OptionalListField struct {
	Items       []Item         `json:"items,omitempty"`
	Dictionary  map[Key]Value  `json:"dictionary,omitempty"`
}
```

### PR DESCRIPTION
**Pull request by ravirajj**
This changes code generation to enable the `omitempty` JSON option for
optional collections to reduce `null` noise in the JSON output.

So,

    struct OptionalListField {
        1: optional list<Item> items
        2: optional map<Key, Value> dictionary
    }

Becomes,

    type OptionalListField struct {
      Items       []Item         `json:"items,omitempty"`
      Dictionary  map[Key]Value  `json:"dictionary,omitempty"`
    }

Encoding `OptionalListField{}` will now produce, `{}` instead of,

    {"items": null, "dictionary": null}

Resolves #330

### CODE REVIEW

File: compile/container.go [hunk 1]
```diff
@@ -93,14 +93,14 @@ func (m *MapSpec) TypeCode() wire.Type {
 }
 
 // ForEachTypeReference for MapSpec
-func (m *MapSpec) ForEachTypeReference(f func(TypeSpec) error) error {
-	if err := f(m.KeySpec); err != nil {
+func (m *MapSpec) ForEachTypeReference(f func(TypeSpec) error) (err error) {
```
Comment (kriskowal) [inline compile/container.go]: The named return is inconsistent with the style below. We rarely use named returns, but when you do, the `return` statement receives no arguments. This also appears to be a style change unrelated to the PR.

File: gen/field.go [hunk 2]
```diff
@@ -147,23 +147,24 @@ func generateTags(f *compile.FieldSpec) (string, error) {
...
 	}
 
+	// If the field name is "-" then it means omit, add no tags
 	if name == "-" {
-		// If the field name is "-" then it means omit, add no tags
 		return t
 	}
 
-	if (isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required && !t.HasOption("omitempty") {
+	// We want to add the "omitempty" json tag if the field is an 'optional" to
```
Comment (kriskowal) [inline gen/field.go]: Nit.

Either:

    is optional

Or

    is an "optional".

Also, mismatch single and double quote.

File: gen/field.go [hunk 3]
```diff
@@ -147,23 +147,24 @@ func generateTags(f *compile.FieldSpec) (string, error) {
...
+	// reduce "null" noise. The "omitempty" json tag specifies that the field
```
Comment (kriskowal) [inline gen/field.go]: capitalize JSON

File: gen/struct_test.go [hunk 4]
```diff
@@ -877,112 +877,131 @@ func TestStructWithDefaults(t *testing.T) {
 
 func TestStructJSON(t *testing.T) {
 	tests := []struct {
-		v interface{}
-		j string
+		v, w interface{}
+		j    string
```
Comment (kriskowal) [inline gen/struct_test.go]: `v` is obviously short for `value`. `j` is also obviously for `JSON`. What does `w` mean? It might be useful to extract the three cases that make use of the `w` field from the table test, stand-alone. Using positional fields for a {k, v} duple is pretty clean.

File: gen/struct_test.go [hunk 5]
```diff
@@ -992,10 +1011,13 @@ func TestStructJSON(t *testing.T) {
 			assert.Equal(t, tt.j, string(encoded))
 		}
 
-		v := reflect.New(reflect.TypeOf(tt.v).Elem()).Interface()
+		u := tt.v
+		if tt.w != nil {
+			u = tt.w
+		}
+		v := reflect.New(reflect.TypeOf(u).Elem()).Interface()
```
Comment (kriskowal) [inline gen/struct_test.go]: Also might be better to use full names here, since they are drifting from being abbreviations of meaningful names. `v` in scope no longer means value, but “type of value”. `w` appears to be an alternate value for purposes of inferring the type from a more complete object.

This would probably all be a lot more clear with the three relevant cases extracted to a clearer test with meaningful names.

File: gen/struct_test.go [hunk 6]
```diff
@@ -877,112 +877,131 @@ func TestStructWithDefaults(t *testing.T) {
...
 			},
-			`{"name":"foo","contact":{"emailAddress":""}}`,
+			j: `{"name":"foo","contact":{"emailAddress":""}}`,
 		},
-		{&tu.EmptyUnion{}, "{}"},
-		{&tu.Document{Pdf: td.PDF("hello")}, `{"pdf":"aGVsbG8="}`},
-		{&tu.Document{Pdf: td.PDF{}}, `{"pdf":""}`},
+		{v: &tu.EmptyUnion{}, j: "{}"},
+		{v: &tu.Document{Pdf: td.PDF("hello")}, j: `{"pdf":"aGVsbG8="}`},
+		{v: &tu.Document{Pdf: td.PDF{}}, w: &tu.Document{Pdf: nil}, j: `{}`},
```
Comment (kriskowal) [inline gen/struct_test.go]: Does the previously written test case continue to pass? Might be better to just replace, remove, or move.

**Comment by abhinav** [comment]
Thanks for the PR! Please address @kriskowal's comments and drop styling changes that are not relevant to this PR. Also, please rebase this change on dev. I've already updated the merge base for the PR to dev.

File: compile/container.go (hunk 1 above)
Comment (ravirajj) [reply compile/container.go]: This change was made to fix the lint error.  I'll update the style.

File: gen/field.go (hunk 3 above)
Comment (ravirajj) [reply gen/field.go]: Done.

File: gen/field.go (hunk 2 above)
Comment (ravirajj) [reply gen/field.go]: Done.

**Comment by ravirajj** [comment]
- Rebased on `dev`.
- Added separate tests for optional empty list/map struct fields.
- `TestStructJSON` already covers optional nil and non-empty list/map struct fields.

File: compile/container.go [hunk 7]
```diff
@@ -97,10 +97,7 @@ func (m *MapSpec) ForEachTypeReference(f func(TypeSpec) error) error {
 	if err := f(m.KeySpec); err != nil {
 		return err
 	}
-	if err := f(m.ValueSpec); err != nil {
-		return err
-	}
-	return nil
+	return f(m.ValueSpec)
```
Comment (kriskowal) [inline compile/container.go]: Again with the unrelated style changes ;)

**Comment by kriskowal** [comment]
Please get a stamp from @abhinav as well. We require two stamps for software networking projects.

File: compile/container.go (hunk 7 above)
Comment (ravirajj) [reply compile/container.go]: I had to make these changes to fix the lint errors!

Comment (abhinav) [reply compile/container.go]: This is a lint error in Go 1.8 but not in 1.9. I changed Travis on this repo
to run for Go 1.9 only. This should no longer be a problem.

**Comment by abhinav** [comment]
Merged. Thanks for the change!
